File: src/Effect.cs
using System;

namespace Effect;

public static class FFI {
    public static Func<object> PureE(object a) => () => a;
    
    public static Func<object> BindE(Func<object> a, Func<object, Func<object>> f) => () => {
        var resA = a();
        return f(resA)();
    };
    
    public static Func<object> UntilE(Func<bool> f) => () => {
        while (!f()) {}
        return null;
    };
    
    public static Func<object> WhileE(Func<bool> f, Func<object> a) => () => {
        while (f()) {
            a();
        }
        return null;
    };
    
    public static Func<object> ForE(long lo, long hi, Func<long, Func<object>> f) => () => {
        for (long i = lo; i < hi; i++) {
            f(i)();
        }
        return null;
    };
    
    public static Func<object> ForeachE(object[] as_, Func<object, Func<object>> f) => () => {
        foreach (var v in as_) {
            f(v)();
        }
        return null;
    };
}

File: src/Effect.fs
let pureE = box (fun (a: obj) -> box (fun _ -> a))
let bindE = box (fun (a: obj) -> box (fun (f: obj) -> box (fun _ -> 
    let a' = a :?> (obj -> obj)
    let f' = f :?> (obj -> obj)
    let a_res = a' null
    let f_res = f' a_res :?> (obj -> obj)
    f_res null
)))

let untilE = box (fun (f: obj) -> box (fun _ ->
    let f' = f :?> (obj -> obj)
    let mutable condition = false
    while not condition do
        condition <- unbox<bool> (f' null)
    null
))

let whileE = box (fun (f: obj) -> box (fun (a: obj) -> box (fun _ ->
    let f' = f :?> (obj -> obj)
    let a' = a :?> (obj -> obj)
    let mutable condition = unbox<bool> (f' null)
    while condition do
        a' null |> ignore
        condition <- unbox<bool> (f' null)
    null
)))

let forE = box (fun (lo: obj) -> box (fun (hi: obj) -> box (fun (f: obj) -> box (fun _ ->
    let f' = f :?> (obj -> obj)
    let l = unbox<int> lo
    let h = unbox<int> hi
    for i = l to h - 1 do
        let step = f' (box i) :?> (obj -> obj)
        step null |> ignore
    null
))))

let foreachE = box (fun (arr: obj) -> box (fun (f: obj) -> box (fun _ ->
    let f' = f :?> (obj -> obj)
    let arr' = unbox<obj[]> arr
    for v in arr' do
        let step = f' v :?> (obj -> obj)
        step null |> ignore
    null
)))
